### PR TITLE
Floatingactionmenu

### PR DESCRIPTION
**Pull request by ErlingMK**
## Added / Fixed
- Added a FloatingActionMenu. Add as Behaviour to Absolute, Relative or ModalityLayouts

## Todo List

- [x] I have tested on an Android device.
- [x] I have tested on an iOS device.
- [ ] I have added unit tests

>*Please add pictures / Gifs if possible*
>*Todo List can be checked by putting a `X` inside the brackets*
>*Remember to update our `wiki` after this PR is merged*

### CODE REVIEW

**Comment by ErlingMK** [comment]
FloatingMenu

File: src/DIPS.Xamarin.UI/Controls/FloatingActionMenu/FloatingActionMenu.xaml.cs [hunk 1]
```diff
@@ -0,0 +1,307 @@
...
+            await ExpandButton.RelRotateTo(180, 250, Easing.CubicInOut);
+            if (isExpanded) m_overLay.IsVisible = false;
+            m_isExpanded = !isExpanded;
+            m_animationComplete = true;
+        }
+
+        private async void ExpandButton_OnClicked(object sender, EventArgs e)
+        {
+            if (m_first) AdjustXPositions();
+            if (m_animationComplete) await AnimateAll(m_isExpanded);
```
Comment (Illedan) [inline src/DIPS.Xamarin.UI/Controls/FloatingActionMenu/FloatingActionMenu.xaml.cs]: Just set ExpandButton.InputTransparent = true before and false after Animate all.

File: src/DIPS.Xamarin.UI/Controls/FloatingActionMenu/FloatingActionMenu.xaml.cs [hunk 2]
```diff
@@ -0,0 +1,307 @@
...
+                if (m_parent is AbsoluteLayout)
+                {
+                    AbsoluteLayout.SetLayoutFlags(child, AbsoluteLayoutFlags.None);
+                    AbsoluteLayout.SetLayoutBounds(
+                        child,
+                        new Rectangle(ExpandButton.X + Size - child.Width, ExpandButton.Y, child.Width, child.Height));
+                }
+                else if (m_parent is RelativeLayout)
+                {
+                    RelativeLayout.SetXConstraint(child, Constraint.Constant(ExpandButton.X + Size - child.Width));
```
Comment (Illedan) [inline src/DIPS.Xamarin.UI/Controls/FloatingActionMenu/FloatingActionMenu.xaml.cs]: Expandbutton is placed on the contentView itself? Did you test this working?

File: src/DIPS.Xamarin.UI/Controls/FloatingActionMenu/FloatingActionMenu.xaml.cs [hunk 3]
```diff
@@ -0,0 +1,307 @@
...
+            else
+            {
+                m_overLay.FadeTo(0);
+            }
+
+            var multiplier = 1;
+
+            foreach (var menuButton in Children)
+            {
+                menuButton.InputTransparent = isExpanded;
```
Comment (Illedan) [inline src/DIPS.Xamarin.UI/Controls/FloatingActionMenu/FloatingActionMenu.xaml.cs]: Set buttons to clickable after all animations?

File: src/DIPS.Xamarin.UI/Controls/FloatingActionMenu/FloatingActionMenuBehaviour.cs [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+            base.OnAttachedTo(bindable);
+
+            m_floatingActionMenu = Content as FloatingActionMenu ?? Content?.Children.First(c => c.GetType() == typeof(FloatingActionMenu)) as FloatingActionMenu;
+            if (!(bindable is ModalityLayout)) m_floatingActionMenu?.AddTo(bindable);
+            bindable.SizeChanged += BindableOnSizeChanged;
+        }
+
+        private void BindableOnSizeChanged(object sender, EventArgs e)
+        {
+            if (sender is ModalityLayout modality) m_floatingActionMenu?.AddTo(modality.relativeLayout);
```
Comment (Illedan) [inline src/DIPS.Xamarin.UI/Controls/FloatingActionMenu/FloatingActionMenuBehaviour.cs]: Add { and } with IFs

File: src/DIPS.Xamarin.UI/Controls/FloatingActionMenu/FloatingActionMenu.xaml.cs [hunk 5]
```diff
@@ -0,0 +1,307 @@
...
+        ///     FontSize of the text in the expand button.
+        /// </summary>
+        public double FontSize
+        {
+            get => (double)GetValue(FontSizeProperty);
+            set => SetValue(FontSizeProperty, value);
+        }
+
+        /// <summary>
+        ///     The Z-position of the control. Is proportional to the Layout it's added to. Values between 0-1.
```
Comment (haavamoa) [inline src/DIPS.Xamarin.UI/Controls/FloatingActionMenu/FloatingActionMenu.xaml.cs]: Is Z position the correct name here? The property is called xConstrain

Comment (haavamoa) [reply src/DIPS.Xamarin.UI/Controls/FloatingActionMenu/FloatingActionMenu.xaml.cs]: And, can it be called XPosition? The comment mentions position

File: src/DIPS.Xamarin.UI/Controls/FloatingActionMenu/FloatingActionMenu.xaml.cs [hunk 6]
```diff
@@ -0,0 +1,307 @@
...
+        ///     FontFamily of the text in the expand button.
+        /// </summary>
+        public string FontFamily
+        {
+            get => (string)GetValue(FontFamilyProperty);
+            set => SetValue(FontFamilyProperty, value);
+        }
+
+        /// <summary>
+        /// </summary>
```
Comment (haavamoa) [inline src/DIPS.Xamarin.UI/Controls/FloatingActionMenu/FloatingActionMenu.xaml.cs]: Add comment

File: src/DIPS.Xamarin.UI/Controls/FloatingActionMenu/FloatingActionMenu.xaml.cs [hunk 7]
```diff
@@ -0,0 +1,307 @@
...
+        public double Size
+        {
+            get => (double)GetValue(SizeProperty);
+            set => SetValue(SizeProperty, value);
+        }
+
+        /// <summary>
+        ///     Set to true to close the menu if its open.
```
Comment (haavamoa) [inline src/DIPS.Xamarin.UI/Controls/FloatingActionMenu/FloatingActionMenu.xaml.cs]: Describe what it is instead of telling what to do

Comment (haavamoa) [reply src/DIPS.Xamarin.UI/Controls/FloatingActionMenu/FloatingActionMenu.xaml.cs]: Can this property be called IsOpen? Remember to flip its usage if you do

File: src/Samples/DIPS.Xamarin.UI.Samples/Controls/FloatingActionMenu/FloatingActionMenuPage.xaml [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+             xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"
+             xmlns:d="http://xamarin.com/schemas/2014/forms/design"
+             xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
+             xmlns:dxui="http://dips.xamarin.ui.com"
+             xmlns:FloatingActionMenu="clr-namespace:DIPS.Xamarin.UI.Samples.Controls.FloatingActionMenu;assembly=DIPS.Xamarin.UI.Samples"
+             mc:Ignorable="d"
+             x:Class="DIPS.Xamarin.UI.Samples.Controls.FloatingActionMenu.FloatingActionMenuPage">
+    <ContentPage.Content>
+        <RelativeLayout>
+            <RelativeLayout.Behaviors>
```
Comment (Illedan) [inline src/Samples/DIPS.Xamarin.UI.Samples/Controls/FloatingActionMenu/FloatingActionMenuPage.xaml]: Do you have an example of how to make this into a reusable component to share in a big project?

File: src/Samples/DIPS.Xamarin.UI.Samples/Controls/FloatingActionMenu/FloatingActionMenuPage.xaml [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+                <dxui:FloatingActionMenuBehaviour>
```
Comment (haavamoa) [inline src/Samples/DIPS.Xamarin.UI.Samples/Controls/FloatingActionMenu/FloatingActionMenuPage.xaml]: Didn't we agree on reusing ModalityLayout?

File: src/DIPS.Xamarin.UI/Controls/FloatingActionMenu/FloatingActionMenuBehaviour.cs [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+using DIPS.Xamarin.UI.Controls.Modality;
+using Xamarin.Forms;
+
+namespace DIPS.Xamarin.UI.Controls.FloatingActionMenu
+{
+    /// <summary>
+    /// 
+    /// </summary>
+    [ContentProperty(nameof(Content))]
+    public class FloatingActionMenuBehaviour : Behavior<Layout>
```
Comment (haavamoa) [inline src/DIPS.Xamarin.UI/Controls/FloatingActionMenu/FloatingActionMenuBehaviour.cs]: Will this work with ModalityLayout? ModalityLayout is not of type layout.

File: src/DIPS.Xamarin.UI/Controls/FloatingActionMenu/FloatingActionMenu.xaml.cs (hunk 2 above)
Comment (ErlingMK) [reply src/DIPS.Xamarin.UI/Controls/FloatingActionMenu/FloatingActionMenu.xaml.cs]: Unsure what you mean?

File: src/Samples/DIPS.Xamarin.UI.Samples/Controls/FloatingActionMenu/FloatingActionMenuPage.xaml (hunk 9 above)
Comment (ErlingMK) [reply src/Samples/DIPS.Xamarin.UI.Samples/Controls/FloatingActionMenu/FloatingActionMenuPage.xaml]: After we talked about the intended use of the ModalityLayout I decided to do it like this instead. It also removes the need to use a modality layout when an absolute layout might make more sense. Now it works for both

File: src/Samples/DIPS.Xamarin.UI.Samples/Controls/FloatingActionMenu/FloatingActionMenuPage.xaml (hunk 8 above)
Comment (ErlingMK) [reply src/Samples/DIPS.Xamarin.UI.Samples/Controls/FloatingActionMenu/FloatingActionMenuPage.xaml]: Yes, I have used it like this in an app

image

File: src/DIPS.Xamarin.UI/Controls/FloatingActionMenu/FloatingActionMenu.xaml.cs (hunk 1 above)
Comment (ErlingMK) [reply src/DIPS.Xamarin.UI/Controls/FloatingActionMenu/FloatingActionMenu.xaml.cs]: Doesn't work if the expand button is clicked quickly enough. The change of input transparancy is apparently not instant

File: src/DIPS.Xamarin.UI/Controls/FloatingActionMenu/FloatingActionMenu.xaml.cs (hunk 2 above)
Comment (Illedan) [reply src/DIPS.Xamarin.UI/Controls/FloatingActionMenu/FloatingActionMenu.xaml.cs]: I just wonder if you verified all regarding RelativeLayout

Comment (ErlingMK) [reply src/DIPS.Xamarin.UI/Controls/FloatingActionMenu/FloatingActionMenu.xaml.cs]: Yes, its working for both Relative and Absolute, and modality

**Comment by haavamoa** [comment]
Can you modify your samples to use the ModalityLayout? We can mention in the wiki post for this component that it can be used without a modalitylayout. 

We want our consumers to stick with using a ModalityLayout to get the best user experience of the overlay, so not using it should be mentioned and not sampled imo

File: src/DIPS.Xamarin.UI/Internal/Xaml/FloatingActionMenu.xaml.cs [hunk 11]
```diff
@@ -0,0 +1,174 @@
...
+        }
+
+        private void HideOverlay()
+        {
+            m_parent?.HideOverlay();
+        }
+
+        private async void ExpandButton_OnClicked(object sender, EventArgs e)
+        {
+            if (m_first)
```
Comment (Illedan) [inline src/DIPS.Xamarin.UI/Internal/Xaml/FloatingActionMenu.xaml.cs]: Move if into the AdjustXPositions

File: src/DIPS.Xamarin.UI/Internal/Xaml/FloatingActionMenu.xaml.cs [hunk 12]
```diff
@@ -0,0 +1,174 @@
...
+            {
+                AdjustXPositions();
+            }
+
+            if (m_animationComplete)
+            {
+                await AnimateAll();
+            }
+            m_behaviour.IsOpen = m_isExpanded;
```
Comment (Illedan) [inline src/DIPS.Xamarin.UI/Internal/Xaml/FloatingActionMenu.xaml.cs]: Isn't it better to just return when `m_animationComplete = true`?
Now the IsExpanded is swapped even though no animation would happend?

**Comment by haavamoa** [comment]
Remember to update wiki @ErlingMK

**Comment by chr0m1ng** [comment]
Is there a way to add an image instead of the text?

**Comment by haavamoa** [comment]
> Is there a way to add an image instead of the text?

We are currently using font icons internally. This means that we set the `Text` property and `FontFamily` to get an icon from our internal icon fonts. 

 But I am pretty sure @ErlingMK could have a look at adding a `Icon` property that will use an `Image` as the source of the icon. As this is not an internal need I can not promiss when we will add it. 😔

**Comment by chr0m1ng** [comment]
> > Is there a way to add an image instead of the text?
> 
> We are currently using font icons internally. This means that we set the `Text` property and `FontFamily` to get an icon from our internal icon fonts.
> 
> But I am pretty sure @ErlingMK could have a look at adding a `Icon` property that will use an `Image` as the source of the icon. As this is not an internal need I can not promiss when we will add it. 😔

Wow that was fast haha. Thanks for answering, I achieved what I wanted using `FontFamily` too, just implemented that, thanks for the tip.

Maybe you guys could make it more explicit on the docs for the future.

BTW: Awesome library

**Comment by haavamoa** [comment]
> > > Is there a way to add an image instead of the text?
> > 
> > 
> > We are currently using font icons internally. This means that we set the `Text` property and `FontFamily` to get an icon from our internal icon fonts.
> > But I am pretty sure @ErlingMK could have a look at adding a `Icon` property that will use an `Image` as the source of the icon. As this is not an internal need I can not promiss when we will add it. 😔
> 
> Wow that was fast haha. Thanks for answering, I achieved what I wanted using `FontFamily` too, just implemented that, thanks for the tip.
> 
> Maybe you guys could make it more explicit on the docs for the future.
> 
> BTW: Awesome library

Good point. We will update doc. 👍

**Comment by haavamoa** [comment]
Issue created : https://github.com/DIPSAS/DIPS.Xamarin.UI/issues/201

**Pull request merged**